pkg/services/ngalert/image/service.go

```diff
@@ -0,0 +1,119 @@
+	"github.com/grafana/grafana/pkg/services/rendering"
+	"github.com/grafana/grafana/pkg/services/screenshot"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+//go:generate mockgen -destination=mock.go -package=image github.com/grafana/grafana/pkg/services/ngalert/image ImageService
+type ImageService interface {
+	// NewImage returns a new image for the alert instance.
+	NewImage(ctx context.Context, r *ngmodels.AlertRule, labels data.Labels) (*store.Image, error)
+}
+
+var (
+	// ErrNoDashboard is returned when the alert rule does not have a dashboard.
+	ErrNoDashboard = errors.New("no dashboard")
+
+	// ErrNoPanel is returned when the alert rule does not have a panel in a dashboard.
+	ErrNoPanel = errors.New("no panel")
+)
+
+const (
+	screenshotTimeout  = 10 * time.Second
+	screenshotCacheTTL = 15 * time.Second
+)
+
+// ScreenshotImageService takes screenshots of the panel for an alert rule and
+// saves the image in the store. The image contains a unique token that can be
+// passed as an annotation or label to the Alertmanager.
+type ScreenshotImageService struct {
+	screenshots screenshot.ScreenshotService
+	store       store.ImageStore
+}
+
+func NewScreenshotImageService(screenshots screenshot.ScreenshotService, store store.ImageStore) ImageService {
+	return &ScreenshotImageService{
+		screenshots: screenshots,
+		store:       store,
+	}
+}
+
+// NewScreenshotImageServiceFromCfg returns a new ScreenshotImageService
+// from the configuration.
+func NewScreenshotImageServiceFromCfg(cfg *setting.Cfg, metrics prometheus.Registerer,
+	db *store.DBstore, ds dashboards.DashboardService, rs rendering.Service) (ImageService, error) {
+	if !cfg.UnifiedAlerting.Screenshots.Enabled {
+		return &ScreenshotImageService{
+			screenshots: &screenshot.ScreenshotUnavailableService{},
+		}, nil
+	}
+
+	s := screenshot.NewBrowserScreenshotService(metrics, ds, rs)
+	if cfg.UnifiedAlerting.Screenshots.UploadExternalImageStorage {
+		u, err := imguploader.NewImageUploader()
+		if err != nil {
+			return nil, fmt.Errorf("failed to initialize uploading screenshot service: %w", err)
+		}
+		s = screenshot.NewUploadingScreenshotService(metrics, s, u)
+	}
+	s = screenshot.NewRateLimitScreenshotService(s, cfg.UnifiedAlerting.Screenshots.MaxConcurrentScreenshots)
+	s = screenshot.NewSingleFlightScreenshotService(s)
+	s = screenshot.NewCachableScreenshotService(metrics, screenshotCacheTTL, s)
+	s = screenshot.NewObservableScreenshotService(metrics, s)
+
+	return &ScreenshotImageService{
+		store:       db,
+		screenshots: s,
+	}, nil
+}
+
+// NewImage returns a screenshot of the panel for the alert rule. It returns
+// ErrNoDashboard if the alert rule does not have a dashboard and ErrNoPanel
+// when the alert rule does not have a panel in a dashboard.
+func (s *ScreenshotImageService) NewImage(ctx context.Context, r *ngmodels.AlertRule, _ data.Labels) (*store.Image, error) {
+	if r.DashboardUID == nil {
+		return nil, ErrNoDashboard
+	}
+	if r.PanelID == nil {
```

Comment:
Do we care if it's set but it's zero? We seem to assume it's nonzero in other places.

Reply:
Zero will result in an invalid request, thus I think we should catch it here.